proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,1189 @@
+
+    module C where
+    bar = 42
+
+    module D where
+    baz 0 = True
+    baz _ = False
+
+
+Now consider compiling ``A`` at stage *R*.
+
+* ``B`` is required at stage *C*, as it is splice imported from ``A`` at *R*.
+* ``C`` is required at stage *R*, as it is quote imported from ``B`` at *C*.
+* ``D`` is required at stage *C*, as it is normally imported from ``B`` at *C*.
+
+Therefore in order to compile ``A`` at *R*, we have performed
+dependency resolution and require ``B`` at *C*, ``C`` at *R* and ``D`` at *C*.
+
+The perhaps curious case is ``D``: is it needed at compile-time or runtime? It
+does not use a splice import, so one could think it is needed at runtime -- but
+here is where the distinction between the import level offset and base stage is
+relevant. ``D`` is only being imported as a dependency of ``B``, which is at *C*
+stage. This makes ``D`` *also* at the *C* stage! Note how ``baz`` is needed
+at compile time just to define ``foo``, which is properly ``splice`` imported.
+
+The levels of all modules in the transitive closure of a ``splice``-imported
+module are offset by -1. Conversely, ``quote`` imports offset the levels by +1,
+thereby making all the levels align correctly.
+
+Implicit Stage Persistence and Stages
+#####################################
+
+Modules using implicit stage persistence place a set of strong requirements on itself and
+immediate dependencies. Consider this example where module ``B`` uses ``ImplicitStagePersistence``::
+
+  module A where { a = 1 :: Int }
+
+  {-# LANGUAGE ExplicitLevelImports #-}
+  {-# LANGUAGE ImplicitStagePersistence #-}
+  module B where
+  import A
+
+  foo = a
+
+  bar = [| foo |]
+
+  {-# LANGUAGE ExplicitLevelImports #-}
+  module C where
+  import splice B
+  c :: Int
+  c = $(bar)
+
+Consider compiling ``C @ R``, when ``bar`` from ``B`` is executed, then
+it will produce a program ``B.foo``. Therefore we will also need ``B @ R``.
+
+How could we determine from the module header that we would require ``B @ R``?
+
+* ``C @ R`` splice imports ``B``, therefore only directly places a requirement on ``B @ C``
+* However, ``B`` enables ``ImplicitStagePersistence``, and therefore is able to persist
+  top-level definitions and definitions defined in ``B`` itself and all its level 0 or level 1 imports. Therefore we
+  determine we also require ``C @ R``.
+
+
+In this example you can observe that the ability to move a variable between
+levels using cross-stage persistence places a strong set of requirements on the
+stages that modules are required at. Implicit stage pesistence makes imported
+identifiers available at all levels, as a consequence, they must also be available
+at all stages. The introduction of the ``ImplicitStagePersistence``
+extension is wholly motivated by the desire to control these requirements in an explicit
+fashion.
+
+ImplicitStagePersistence, Stages and TemplateHaskellQuotes
+##########################################################
```

Comment:
Is the behaviour described in this section supposed to be part of the specification?

Reply:
The logic behind the approach we've taken is that the specification deals with which programs are part of the language / accepted by the compiler (hence it deals exclusively with *level*), whereas *stage* is supposed to be an implementation concept (so it affects what things get compiled when, but not whether programs are accepted at all). Thus we don't nail down exactly how the build system works in the spec, but of course the examples needed to motivate the design are expressed in terms of implications for the implementation.